[rocshmem] Add Dockerfile and doc of how to use (#4790)
ROCm/rocm-systems: projects/rocshmem/docs/docker.rst

.. meta::
   :description: Instructions for building and running rocSHMEM using Docker containers.
   :keywords: rocSHMEM, ROCm, Docker, container, build, run, GDA, IPC, RO

.. _docker-rocshmem:

------------------------------------------
Using the rocSHMEM Docker container
------------------------------------------

> rocSHMEM provides a Dockerfile in ``scripts/build_configs/Dockerfile`` that builds a

Reviewer: ```suggestion
rocSHMEM provides a Dockerfile in ``docker/Dockerfile.ubuntu`` that builds a
```

development container with all backends, MPI support, and optional PR testing.

Building the container image
----------------------------

> Build the default image from the ``scripts/build_configs/`` directory:

Reviewer: ```suggestion
Build the default image from the ``docker/`` directory:
```

.. code-block:: bash

>    cd scripts/build_configs

Reviewer: ```suggestion
   cd docker
```

>    docker build --tag $USER/rocshmem .

Reviewer: ```suggestion
   docker build --tag $USER/rocshmem -f Dockerfile.ubuntu .
```

I assume the same for the other docker build commands in this doc

Build arguments
^^^^^^^^^^^^^^^

``PR_NUM``
   Build and test a specific pull request branch alongside the ``develop`` branch.

   .. code-block:: bash

>       docker build --build-arg PR_NUM=4574 --tag $USER/rocshmem-pr4574 .

Reviewer: ```suggestion
      docker build --build-arg PR_NUM=4574 --tag $USER/rocshmem-pr4574 -f Dockerfile.ubuntu .
```

   -v /etc/libibverbs.d/ionic.driver:/etc/libibverbs.d/ionic.driver

Launching persistent multi-node containers
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The default container command starts ``rocshmem_info`` then launches an SSH daemon,
which enables ``mpiexec`` across containers. Use a job launcher (``srun``,
``pdsh``, Kubernetes, etc.) to deploy containers on multiple nodes:

.. code-block:: bash

   srun -N2 --ntasks-per-node=1 --gpus=8 \
     docker run -d --rm \
       --shm-size 64G --network host \
       --device /dev/dri --device /dev/kfd --device /dev/infiniband \
       --ipc host --group-add video --cap-add SYS_PTRACE \
       --security-opt seccomp=unconfined --privileged \
       --name $USER-rocshmem $USER/rocshmem

.. note::

   All containers must use the same Docker image so they share the same SSH key
   generated during the build. Alternatively, import an invariant SSH key pair by
   uncommenting the ``COPY id_ed25519`` lines in the Dockerfile.

Running tests
^^^^^^^^^^^^^

Run a single functional test (for example, ``rocshmem_waveput_nbi``) with 16
workgroups and 256 threads per workgroup inside persistent containers:

.. code-block:: bash

>    srun -n1 docker exec -it -e ROCSHMEM_DEBUG_LEVEL=ENV \

Reviewer: Maybe note that it requires an active container with the name `$USER-rocshmem` for this command to work.

Seems obvious to me but not sure if it is for everyone
